### ISSUE TITLE
Switch to a custom subclass of ServerRequest

### ISSUE THREAD

**Issue by tobyzerner** [flarum/core#869]
Currently to get Flarum-specific attributes on the request you need to go:

``` php
$request->getAttribute('actor'); // get the actor
$request->getAttribute('session'); // get the session
// etc.
```

Let's solidify this into an actual API by subclassing Zend\Diactoros\ServerRequest:

``` php
namespace Flarum\Http;

class ServerRequest extends \Zend\Diactoros\ServerRequest
{
    public function getActor();
    public function withActor();
    public function getSession();
    public function withSession();
    // etc.
}
```

We'll then convert a SeverRequestInterface object into our subclass at some point in the request lifecycle. And we'll typehint all of our methods with our subclass rather than ServerRequestInterface, since they do actually depend on some Flarum-specific attributes being present.

**Comment by franzliedke** [comment]
I was going to complain that this breaks compatibility with other PSR-7 packages, but it doesn't, so I like the idea. :)

I suppose we'd do another middleware that takes the request and wraps it with our custom class, correct?

**Comment by askvortsov1** [comment]
@franzliedke I made a PR with a subclass at https://github.com/flarum/core/pull/2048/files. I'm currently methods for actor, session, bypassCsrfToken, bypassFloodgate, and locale. Anything I missed / extra?

**Comment by franzliedke** [comment]
Hmm, to be honest, I don't think I am too sold on this idea anymore. For example, typehinting our custom class in the signature of controllers would break compatibility with the PSR-15 request handler interface.

If the primary goal is better type hinting (IDE support etc.) and a bit more safety, then I would say an API such as `Actor::fromRequest()` would give us the same benefits as well, and a nice place to bail out if something is completely wrong.

**Comment by askvortsov1** [comment]
I'd be fine with this. Is this something we need in 0.1? Adding that API shouldn't break backwards compatibility as far as I know.

**Comment by luceos** [comment]
> For example, typehinting our custom class in the signature of controllers would break compatibility with the PSR-15 request handler interface.

PSR-15 defines an interface, our implementation could still implement that. Having said that, I think I still agree with not subclassing anything, but using a different class for retrieval.

**Comment by askvortsov1** [comment]
Let's go with this then, but I'd like to contain all these in a single `Flarum\Http\RequestUtil` with `setActor`, `getActor`, etc methods that take a request as a single argument. We could also have `setActorOnRequest` and `getActorFromRequest`, but that's a bit too wordy imo

**Comment by askvortsov1** [comment]
@franzliedke An alternative idea that came up was registering the current user with the container. How would this square with your stance on https://github.com/flarum/core/issues/2307, where you mentioned auth as one of the issues? The current user is tied to the request by nature of HTTP, but it's not necessarily a per-request concept (if we used websockets instead of HTTP, there wouldn't necessarily be a connection. On the other hand, it's not quite application state.

**Comment by askvortsov1** [comment]
The next thing to decide is which attributes we want to provide abstraction for. `actor` seems like the most necessary, but do we want to do others? For instance:

actor
session
locale
route name

**Comment by askvortsov1** [comment]
I've thought a bit more about this. I don't think typehinting itself warrants this system; to me, the biggest benefits come from encapsulation of the current user (enabling #1847).

I would propose we go with a version of https://github.com/flarum/core/pull/2449 that only supports getting the actor. I would also propose that we consider making this a static method of the `User` model instead of a whole new class.

**Comment by askvortsov1** [comment]
Copying a critical convo from #2703 so it doesn't get lost.

luceos said:

> Now that I see this code I am not sure I like it. This approach also goes against our use other conventions. I'm increasingly leaning towards creating an Actor interface (applied to User) and then binding the authenticated actor into ioc from the middleware. Thoughts?

I replied:

> There's 2 main reasons:
>
> - Typehinting Flarum-specific request stuff (I don't think this is a strong enough reason)
> - Encapsulating setting and retrieval of the actor on requests to make #1847 simpler. This, I think, is valid.
>
> I don't really like any of the proposed solutions. A custom subclass is an unnecessary layer. Methods on User are unrelated responsibility. And generally speaking, Util classes are indicators of bad design. But I'm not sure there's a better option. Maybe we could call the class Actor instead of RequestUtil?
>
> As an alternative, we could solve #1847 via the attached PR, and close this issue entirely. The actorReference attribute would be considered private API, but THAT is messy because custom auth middleware added by extensions wouldn't know to use actorReference. Or we could switch to $request->get('actorReference')->getActor() and $request->get('actorReference')->setActor(), but that's very verbose, and would be an impactful breaking change.

With this in mind, I'm closing #2703 and reopening #2449
